src/functions/env.ts

```diff
@@ -264,44 +264,67 @@ export function writeUserEnvs(toWrite: Record<string, string>, envOpts: UserEnvs
   if (Object.keys(toWrite).length === 0) {
     return;
   }
-
   const { functionsSource, projectId, projectAlias, isEmulator } = envOpts;
-  const envFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
-  const projectScopedFileName = `.env.${projectId}`;
 
-  const projectScopedFileExists = envFiles.includes(projectScopedFileName);
-  if (!projectScopedFileExists) {
-    createEnvFile(envOpts);
+  const allEnvFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
+  const targetEnvFile = envOpts.isEmulator
+    ? FUNCTIONS_EMULATOR_DOTENV
+    : `.env.${envOpts.projectId}`;
+  const targetEnvFileExists = allEnvFiles.includes(targetEnvFile);
+  if (!targetEnvFileExists) {
+    fs.writeFileSync(path.join(envOpts.functionsSource, targetEnvFile), "", { flag: "wx" });
+    if (!isEmulator) {
+      logBullet(
+        clc.yellow(clc.bold("functions: ")) +
+          `Created new local file ${targetEnvFile} to store param values. We suggest explicitly adding or excluding this file from version control.`
+      );
+    }
   }
 
-  const currentEnvs = loadUserEnvs(envOpts);
   for (const k of Object.keys(toWrite)) {
     validateKey(k);
-    if (currentEnvs.hasOwnProperty(k)) {
+    if (isAlreadyDefinedWrite(k, envOpts)) {
       throw new FirebaseError(
         `Attempted to write param-defined key ${k} to .env files, but it was already defined.`
       );
     }
   }
 
   logBullet(
-    clc.cyan(clc.bold("functions: ")) +
-      `Writing new parameter values to disk: ${projectScopedFileName}`
+    clc.cyan(clc.bold("functions: ")) + `Writing new parameter values to disk: ${targetEnvFile}`
   );
   for (const k of Object.keys(toWrite)) {
     fs.appendFileSync(
-      path.join(functionsSource, projectScopedFileName),
+      path.join(functionsSource, targetEnvFile),
       formatUserEnvForWrite(k, toWrite[k])
     );
   }
 }
 
-function createEnvFile(envOpts: UserEnvsOpts): string {
-  const fileToWrite = envOpts.isEmulator ? FUNCTIONS_EMULATOR_DOTENV : `.env.${envOpts.projectId}`;
-  logger.debug(`Creating ${fileToWrite}...`);
+/**
+ * Returns whether we should error on trying to write a key because it's already
+ * defined in the .env fields. This seems like a simple presence, check, but...
+ *
+ * For emulator deploys, it's legal to write a key to .env.local even if it's
+ * already defined in .env.projectId. This is a special case designed to follow
+ * the principle of least surprise for emulator users.
+ */
+export function isAlreadyDefinedWrite(key: string, envOpts: UserEnvsOpts): boolean {
+  const definedInEnv = loadUserEnvs(envOpts).hasOwnProperty(key);
```

Comment:
This method is called in a loop. Are we really doing up to 4 file reads for every environment variable?

Reply:
Fixed, though I have some serious concerns about the readability of my fix.